packages/cli/src/modules/mcp/mcp-oauth-token.service.ts

```diff
@@ -0,0 +1,210 @@
+import { Logger } from '@n8n/backend-common';
+import { User, UserRepository } from '@n8n/db';
+import { Service } from '@n8n/di';
+import { randomBytes, randomUUID } from 'node:crypto';
+
+import { JwtService } from '@/services/jwt.service';
+
+import { AccessTokenRepository } from './oauth-access-token.repository';
+import { RefreshTokenRepository } from './oauth-refresh-token.repository';
+
+/**
+ * Manages OAuth 2.1 token lifecycle for MCP server
+ * Generates, validates, rotates, and revokes access and refresh tokens
+ */
+@Service()
+export class McpOAuthTokenService {
+	private readonly MCP_AUDIENCE = 'mcp-server-api';
+	private readonly ACCESS_TOKEN_EXPIRY_SECONDS = 3600; // 1 hour
+	private readonly REFRESH_TOKEN_EXPIRY_MS = 30 * 24 * 3600 * 1000; // 30 days
+
+	constructor(
+		private readonly logger: Logger,
+		private readonly jwtService: JwtService,
+		private readonly userRepository: UserRepository,
+		private readonly accessTokenRepository: AccessTokenRepository,
+		private readonly refreshTokenRepository: RefreshTokenRepository,
+	) {}
+
+	generateTokenPair(
+		userId: string,
+		clientId: string,
+	): { accessToken: string; refreshToken: string } {
+		const accessToken = this.jwtService.sign({
+			sub: userId,
+			aud: this.MCP_AUDIENCE,
+			client_id: clientId,
+			jti: randomUUID(),
+			iat: Math.floor(Date.now() / 1000),
+			exp: Math.floor(Date.now() / 1000) + this.ACCESS_TOKEN_EXPIRY_SECONDS,
+			meta: {
+				isOAuth: true,
+			},
+		});
+
+		const refreshToken = randomBytes(32).toString('hex');
+
+		return { accessToken, refreshToken };
+	}
+
+	async saveTokenPair(
+		accessToken: string,
+		refreshToken: string,
+		clientId: string,
+		userId: string,
+	): Promise<void> {
+		await this.accessTokenRepository.manager.transaction(async (transactionManager) => {
+			await transactionManager.save(this.accessTokenRepository.target, {
+				token: accessToken,
+				clientId,
+				userId,
+			});
+
+			await transactionManager.save(this.refreshTokenRepository.target, {
+				token: refreshToken,
+				clientId,
+				userId,
+				expiresAt: Date.now() + this.REFRESH_TOKEN_EXPIRY_MS,
+			});
+		});
+	}
+
+	async validateAndRotateRefreshToken(
+		refreshToken: string,
+		clientId: string,
+	): Promise<OAuthTokens> {
+		const refreshTokenRecord = await this.refreshTokenRepository.findOne({
+			where: {
+				token: refreshToken,
+				clientId,
+			},
+		});
+
+		if (!refreshTokenRecord) {
+			throw new Error('Invalid refresh token');
+		}
+
+		if (refreshTokenRecord.expiresAt < Date.now()) {
+			await this.refreshTokenRepository.remove(refreshTokenRecord);
+			throw new Error('Refresh token expired');
+		}
+
+		const { accessToken, refreshToken: newRefreshToken } = this.generateTokenPair(
+			refreshTokenRecord.userId,
+			clientId,
+		);
+
+		await this.refreshTokenRepository.remove(refreshTokenRecord);
+
+		await this.saveTokenPair(accessToken, newRefreshToken, clientId, refreshTokenRecord.userId);
```

Comment:
This should ideally happen in a transaction or we can end up in a state where the refersh token was removed but we didn't create new one